awx/api/serializers.py

```diff
@@ -2201,6 +2201,89 @@ def create(self, validated_data):
         return return_data
 
 
+class BulkHostDeleteSerializer(serializers.Serializer):
+    hosts = serializers.ListField(
+        allow_empty=False,
+        max_length=250,
+        write_only=True,
+        help_text=_('List of hosts ids to be deleted, e.g. [105, 130, 131, 200]'),
+    )
+
+    class Meta:
+        model = Host
+        fields = 'hosts'
+        read_only_fields = ()
+
+    def validate(self, attrs):
+        request = self.context.get('request', None)
+
+        # Validating the number of hosts to be deleted
+        if len(attrs['hosts']) > settings.BULK_HOST_MAX_DELETE:
+            raise serializers.ValidationError(
+                {
+                    "ERROR": 'Number of hosts exceeds system setting BULK_HOST_MAX_DELETE',
+                    "BULK_HOST_MAX_DELETE": settings.BULK_HOST_MAX_DELETE,
+                    "Hosts_count": len(attrs['hosts']),
+                }
+            )
+
+        # Getting list of all host objects, filtered by the list of the hosts to delete
+        attrs['in_inv'] = Host.objects.get_queryset().filter(pk__in=attrs['hosts'])
+
+        # Converting the queryset data in a dict. to reduce the number of queries when
+        # manipulating the data
+        hosts_data = attrs['in_inv'].values()
+
+        if len(attrs['in_inv']) == 0:
+            raise serializers.ValidationError({"ERROR": "There are no hosts to delete"})
+
+        if len(attrs['in_inv']) < len(attrs['hosts']):
+            hosts_exists = [host['id'] for host in hosts_data]
+            raise serializers.ValidationError({"ERROR": "Not all hosts are exists", "hosts": list(set(attrs['hosts']).difference(hosts_exists))})
+
+        # Getting all inventories that the hosts can be in
+        inv_list = list(set([host['inventory_id'] for host in hosts_data]))
+
+        errors = {"ERRORS": "Cannot delete host(s) from inventory", "Inventory": dict()}
+        for inv in Inventory.objects.get_queryset().filter(pk__in=inv_list):
+            if inv.kind != '':
+                errors["Inventory"][inv.id] = {
+                    "Inventory_name": inv.name,
+                    "reason": "Hosts can only be deleted from manual inventories (not smart or constructed types).",
+                }
+            if request and not request.user.is_superuser:
+                if request.user not in inv.admin_role:
```

Comment:
Take a look at https://github.com/ansible/awx/blob/87f12a569887e6ab406304941e2ae4b9884bab18/awx/api/serializers.py#L4873


This is more along lines of what we need to do than what we do in bulk host create, where we only need to check 1 inventory.

Don't necessarily need to take whole function since we are only dealing with inventories.
Example:
```              
allowed_invs =set(Inventory.accessible_objects(request.user, 'admin_role').filter(pk__in=[1,4], kind='').values_list('id', flat=True))
not_allowed_invs = set(inv_list) - allowed_invs
```

Reply:
Since I need to loop over a list of inventories to check permission and kind of inventory, I think that this method is good.